### PR TITLE
Migrate FunctionIO tests to ZIO Tests (#1647)

### PR DESCRIPTION
**Pull request by kapke**

### CODE REVIEW

File: core-tests/shared/src/test/scala/zio/FunctionIOSpec.scala [hunk 1]
```diff
@@ -1,202 +1,143 @@
...
+            assertM(add1.zipWith(mul2)(_ -> _).run(6), equalTo(7 -> 12))
+          },
+          testM("`&&&` zips the output of two effectful functions and returns a tuple of their result") {
+            assertM((add1 &&& mul2).run(6), equalTo(7 -> 12))
+          },
+          testM("`|||` computes two effectful functions left and right from from an Either input") {
+            for {
+              l1 <- (add1 ||| mul2).run(Left(25))
+              r1 <- (add1 ||| mul2).run(Right(25))
+            } yield (assert(l1, equalTo(26))
```
Comment (adamgfraser) [inline core-tests/shared/src/test/scala/zio/FunctionIOSpec.scala]: You can get rid of the outer parentheses here and in other tests where you are composing assertions.

Comment (adamgfraser) [reply core-tests/shared/src/test/scala/zio/FunctionIOSpec.scala]: Just put the `&&` at the end of each line.

File: core-tests/shared/src/test/scala/zio/FunctionIOSpec.scala [hunk 2]
```diff
@@ -1,202 +1,143 @@
...
+            assertM(
+              FunctionIO.fail[String]("error").run(1).either,
+              isLeft(equalTo("error"))
+            )
+          },
+          testM("`effect` can translate an Exception to an error") {
+            assertM(thrower.run(9).either, isLeft(equalTo("error")))
+          },
+          testM("`ignore` ignores a effect failure") {
+            assertM(thrower.run(9).ignore, equalTo(()))
```
Comment (adamgfraser) [inline core-tests/shared/src/test/scala/zio/FunctionIOSpec.scala]: You can also use `isUnit` here.

File: core-tests/shared/src/test/scala/zio/FunctionIOSpec.scala [hunk 3]
```diff
@@ -1,202 +1,143 @@
...
-     `ifThenElse` check a pure condition if it is true then computes an effectful function `then0` else computes `else0` $e14b
-     `whileDo` take a condition and run the body until the condition will be  false with impure function $e15a
-     `whileDo` take a condition and run the body until the condition will be  false with pure function $e15b
-     `_1` extracts out the first element of a tuple $e16
-     `_2` extracts out the second element of a tuple $e17
-     `fail` returns a failure  $e18a
-     `effect` can translate an Exception to an error  $e18b
-     `ignore` ignores a effect failure $e19
-    """
+object FunctionIOSpecUtils {
```
Comment (adamgfraser) [inline core-tests/shared/src/test/scala/zio/FunctionIOSpec.scala]: I would move this down to the bottom to mirror the other Specs.

File: core-tests/shared/src/test/scala/zio/FunctionIOSpec.scala [hunk 4]
```diff
@@ -1,202 +1,143 @@
 package zio
 
-import zio.FunctionIO._
+import zio.FunctionIO.{ fail => _, _ }
```
Comment (adamgfraser) [inline core-tests/shared/src/test/scala/zio/FunctionIOSpec.scala]: Any reason you're hiding this? Doesn't look like there is any import conflict and you use it later.

Comment (kapke) [reply core-tests/shared/src/test/scala/zio/FunctionIOSpec.scala]: There was a conflict earlier. I Forgot to simplify this import.

**Pull request merged**